### ISSUE TITLE
PrimeQA NAACL Release : ConfCaliberation

### ISSUE THREAD

**Issue by jdpsen** [primeqa/primeqa#91]
_jaydesen created the following on Mar 21:_  

 Confidence Caliberation module needs to be part of PrimeQA release at NAACL. 
 

_Original issue: https://github.ibm.com/ai-foundation/PrimeQA/issues/63_

**Comment by jdpsen** [comment]
_On Mar 22, zhangr commented:_
#### Features: ####

| Feature | Name | Notes |
| - | - | - |
| 1 | raw_short_answer_score | |
|2 | normalized_short_answer_score|  normalized short answer score with temperature scaling|
| 3 | average_normalized_short_answer_score| average of normalized short score across n-best list|
|4 |cls_span_score | |
| 5 |span_start_score | |
| 6 | span_end_score |  |
| 7 | short_answer_type_score | |
| 8 | no_answer_type_score | |
| 9 | average_minimum_risk_f1 | average of f1 score across n-best list using top-1 answer as reference|
|10 |nbest_has_short_answer| 0/1 indicating if the type of an answer in best-list is short answer|
| 11 | top1_has_short_answer| 0/1 indicating if the type of top-1 answer is short answer |
| 12 |stdev_span_start_score_from_dropout | standard deviation of span start scores obtained by applying random masking to LM output |
| 13 |stdev_span_end_score_from_dropout | standard deviation of span end scores obtained by applying random masking to LM output|

Working on Colbert style features to predict if a question is answerable given context.

**Comment by jdpsen** [comment]
_On Mar 22, zhangr commented:_
#### Model: ####
MLPClassifier (better than regression models)

#### Performance metric: ####
1. F1
2. ECE
<img width="446" alt="ece" src="https://media.github.ibm.com/user/23104/files/cac9c300-a979-11ec-8812-fdfce8e442ac">

**Comment by jdpsen** [comment]
_On Mar 22, zhangr commented:_
#### Experiment: ####

1. Use NQ dev set, about 8000 examples, for training and test.
2. NQ dev set is split into 5 subsets of equal size. Confidence model (MLP classifier) trained on 1st subset, and tested on the other 4 sunsets.

| NQ | F1 | ECE | Notes |
|- | - | - | - |
|No Calibration | 58.48 | 0.3516 | Conf._score = (short_answer_score - min_score) / (max_score - min_score) |
|Temperature scaling | 51.43 | 0.1039 | |
|MLP_C w/o dropout features | 58.09 | 0.0291 | 11 features |
|MLP_C w/ LM output dropout features | 58.32 | 0.0246 | 13 features | 


##### temperature scaling #####

<img width="598" alt="ts" src="https://media.github.ibm.com/user/23104/files/fa7aca00-a97d-11ec-9d90-e835c2acc282">

**Comment by jdpsen** [comment]
_On Mar 22, jaydesen commented:_
conf. caliberation should work on multi lingual tydi.  Similar F1 and  ECE should be acceptable.

**Comment by jdpsen** [comment]
_On Mar 29, jaydesen commented:_
colbert features seem to help, extend it to Tydi data.

**Comment by jdpsen** [comment]
_On Mar 29, zhangr commented:_
Experiment on ColBert features: Cosine distance between question and passage tokens


| NQ | F1 | ECE | Notes |
|- | - | - | - |
|No Calibration | 58.48 | 0.3516 | Conf._score = (short_answer_score - min_score) / (max_score - min_score) |
|Temperature scaling | 51.43 | 0.1039 | |
|MLP_C w/o dropout features | 58.09 | 0.0291 | 11 features |
|MLP_C w/ LM output dropout features | 58.32 | 0.0246 | 13 features | 
|+ ColBert features | 58.52 | 0.0283 | passage-to-question |

**Comment by jdpsen** [comment]
_On Apr 05, jaydesen commented:_
tydi experiment is in progres with all the et of features.

**Comment by jdpsen** [comment]
_On Apr 07, zhangr commented:_
#### Experiment confidence model on TyDi (English) ####
Training examples: 330
Testing examples: 800

| TyDi | F1 | ECE | Notes |
|- | - | - | - |
|No Calibration | 66.05 | 0.2932 |  | 
|Basic Features | 65.38 | 0.0592 | 11 features extracted from system output | 
|+ Dropout features | 65.87 | 0.0610 | 13 features | 
|+ ColBert feature | 66.37 | 0.0617 | 14 features |
|NQ conf. model | 65.43| 0.0971| zero-short, 14 features |  
|Haystack | | | zero-short |

**Comment by jdpsen** [comment]
_On Apr 13, zhangr commented:_
1. create confidence model training/eval pipeline
run_confidence.py
input:
   mrc model
   dataset (tydiqa)
   prediction file
output:
   confidence model

2. enhance mrc pipeline to use confidence scorer
a. add confidence task_head (extractive mrc head + confidence output)
b. extend mrc output to include confidence features
c. add confidence scorer to scorer module
d. change interface to call confidence scorer in decoding

class ExtractiveQAModelOutput(ModelOutput):
    #         # (loss), start_logits, end_logits, target_type_logits, (hidden_states), (attentions)
    loss: Optional[torch.FloatTensor] = None
    start_logits: torch.FloatTensor = None
    end_logits: torch.FloatTensor = None
    target_type_logits: torch.FloatTensor = None
    hidden_states: Optional[Tuple[torch.FloatTensor]] = None
    attentions: Optional[Tuple[torch.FloatTensor]] = None

**Comment by jdpsen** [comment]
duplicate #31

**Comment by jdpsen** [comment]
This is complete.
